keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2868 @@
+
+### Compliance
+
+In-process admission control has fundamental advantages over webhooks: it is far
+safer to use in a "fail closed" mode because it removes the network as a
+possible failure domain. With webhooks, using "fail closed" can negatively
+impact cluster availability. But "fail closed" is very valuable when enforcing
+compliance (and security). We intend to prioritize capabilities that make "fail
+closed" a safe mode of operation. As a concrete example, only allowing CEL
+expressions that pass compilation and type checking significantly reduces the
+opportunities for runtime errors.
+
+Also, making it possible (and convenient) to declare "zero trust" policies is
+important to compliance. By "zero trust", we mean policy rules that apply the
+principle of least privilege to newly created resources (e.g. namespaces) where
+the policy is initially set to the most restrictive state and can be made less
+restrictive via configuration.
+
+## Proposal
+
+Introduce a new `ValidatingAdmissionPolicy` kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to the match rules of admission webhooks, RBAC, priority & fairness and Audit)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  but with access to the data in
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Version conversion support (similar to admission webhook's MatchPolicy)
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+failure policies, type safety, advanced matching rules, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
```

Comment:
Since we require ValidatingAdmissionPolicy, PolicyBinding, parameter CRDs and param configuration to be in place for a complex validation, is there a specify order we required(besides CRD has to be in place before a CR created)? Like Policy configuration has to be in place before the policy definition? I heard "we don't want to force a specific order" and not quite convinced. Also curious When to process the validation and check if all the dependencies in place if we don't ask for order?

Reply:
Thanks for the offline sync @jpbetz . 
Considering the kubectl apply behavior, not forcing a specific order make sense and we have to handle the violation properly. The scenarios would be:
- PolicyDefinition is in place
  - PolicyBinding before param: Needs to be handled properly
  - Policy param before binding: ok
- PolicyDefinition is not in place, it does not hurt in either order 
I guess it is ok to leave bunch of PolicyDefinition there without validation process running.
Since new API is introduced with the feature, it might worth to think about feature support, conversions, backward capability, upgrade/rollout carefully before going to beta.

Reply:
The main ordering problem is when a binding is created before the param CR it requires. If we did check for existence of the param CR and reject the binding write if it is missing, then we would solve the ordering problems but also impose ordering between resources and have to deal with the consequences to `kubectl apply -f <directory-full-of-yaml>`.

I'd prefer to keep with the mitigations mentioned in the KEP and not require ordering, but instead handle broken binding objects in a consistent way.

Good news is that higher level orchestration (e.g. policy engines) can make sure to order things in safe ways so long as we give them the right primitives.

Reply:
Yeah in general it's better to let e.g. controllers retry a few times than to force users to create objects exactly in the right order. In this case it does seem like the admin should get some feedback that the policy is actually being enforced instead of disabled due to configuration error. I think that's something to add to the kep another day if the Audit / Warn settings aren't sufficient.